plugins/mock/devicepluginmock.cpp

```diff
@@ -196,295 +274,493 @@ void DevicePluginMock::startMonitoringAutoDevices()
         if (!m_pushbuttonPressed) {
             qCDebug(dcMockDevice) << "PushButton not pressed yet!";
-            return Device::DeviceSetupStatusFailure;
+            info->finish(Device::DeviceErrorAuthenticationFailure, QT_TR_NOOP("The push button has not been pressed."));
+            return;
         }
 
-        m_pairingId = pairingTransactionId;
-        QTimer::singleShot(1000, this, SLOT(onPushButtonPairingFinished()));
-        return Device::DeviceSetupStatusAsync;
-    } else if (deviceClassId == mockDisplayPinDeviceClassId) {
+        QTimer::singleShot(1000, this, [info](){
+            info->finish(Device::DeviceErrorNoError);
+        });
+        return;
+    }
+
+    if (info->deviceClassId() == mockDisplayPinDeviceClassId) {
         if (secret != "243681") {
             qCWarning(dcMockDevice) << "Invalid pin:" << secret;
-            return Device::DeviceSetupStatusFailure;
+            info->finish(Device::DeviceErrorAuthenticationFailure, QT_TR_NOOP("Invalid PIN!"));
+            return;
         }
-        m_pairingId = pairingTransactionId;
-        QTimer::singleShot(500, this, SLOT(onDisplayPinPairingFinished()));
-        return Device::DeviceSetupStatusAsync;
+        QTimer::singleShot(500, this, [info](){
+            qCDebug(dcMockDevice()) << "Pairing finished.";
+            info->finish(Device::DeviceErrorNoError);
+        });
+        return;
     }
 
-    qCWarning(dcMockDevice) << "Invalid deviceclassId -> no pairing possible with this device";
-    return Device::DeviceSetupStatusFailure;
-}
+    if (info->deviceClassId() == mockUserAndPassDeviceClassId) {
+        qCDebug(dcMockDevice()) << "Credentials received:" << username << secret;
+        if (username == "user" && secret == "password") {
+            info->finish(Device::DeviceErrorNoError);
+            return;
+        } else {
+            info->finish(Device::DeviceErrorAuthenticationFailure, QT_TR_NOOP("Wrong username or password"));
+            return;
+        }
+    }
 
-Device::DeviceError DevicePluginMock::displayPin(const PairingTransactionId &pairingTransactionId, const DeviceDescriptor &deviceDescriptor)
-{
-    Q_UNUSED(pairingTransactionId)
-    Q_UNUSED(deviceDescriptor)
 
-    qCDebug(dcMockDevice) << QString(tr("Display pin!! The pin is 243681"));
+    if (info->deviceClassId() == mockOAuthSonosDeviceClassId) {
+        qCDebug(dcMockDevice()) << "Secret is" << secret;
+        QUrl url(secret);
+        QUrlQuery query(url);
+        qCDebug(dcMockDevice()) << "Acess code is:" << query.queryItemValue("code");
+
+        QString accessCode = query.queryItemValue("code");
+
+        // Obtaining access token
+        url = QUrl("https://api.sonos.com/login/v3/oauth/access");
+        query.clear();
+        query.addQueryItem("grant_type", "authorization_code");
+        query.addQueryItem("code", accessCode);
+        query.addQueryItem("redirect_uri", "https%3A%2F%2F127.0.0.1%3A8888");
```

Comment:
You could use urlEncoded() here for better reading...

Reply:
fixed